mantidproject/mantid
# Convert slice viewer to use bin centres

**Wait for #16922 to be merged before testing this ticket**

The slice viewer currently uses bin edges for histogram workspaces. This is confusing and inconsistent. We should only allow the user to choose slice points which are the centre of histogram bins.

**To test:**
The slice viewer should now only allow you to use the centre (as opposed to the edge) of bins as the slice point on a MDHistoWorkspace.

Two known past issues are #15260 and #16860. As part of the testing for this PR you should follow the instructions below to check both of these problems are mutually resolved:
## Part 1

This checks the problem reported in #16860 is resolved.
1. Run the script below to create an MDWorkspace and open the slice viewer.

``` python
# Lets create a fake MDWorkspace
mdworkspace = CreateMDWorkspace(Dimensions=3, Extents='-10,10,-10,10,-10,10', Names='H,K,L', Units='U,U,U')
# Add A Peak at H=0, K=0, L=1
FakeMDEventData(InputWorkspace=mdworkspace, PeakParams='100000,0,0,0,1')
# Add another peak at H=2, K=2, L=0
FakeMDEventData(InputWorkspace=mdworkspace, PeakParams='100000,2,2,0,1')
# Open SliceViewer and set the slice point to K=1, half-way between the two Peaks.
plotSlice(mdworkspace, xydim=['H','L'], slicepoint=[0,1,0]) 
```
1. Dynamically rebin the workspace with the parameters slice point = 1.0, thickness = 1.0
2. Open the rebinned workspace in the slice viewer. Hover the mouse over the 2D plot and the UI should show that the value of K is 1.0 (consistent with our original slice point).
## Part 2

This checks the problem reported in #15260 is resolved.
1. Run the script below to create an MDWorkspace

``` python
# Create a 3D event workspace to test with
ws_mdew = CreateMDWorkspace(Dimensions='4', Extents='-2,2,-2,2,-2,2,-2,2', Names='h,k,l,E',
                       Units='rlu,rlu,rlu,E', SplitInto='12')
FakeMDEventData(ws_mdew, UniformParams='10000', RandomizeSignal='1')
ws_mdew = SliceMD(ws_mdew,AlignedDim0='h,-2,2,12',AlignedDim1='k,-2,2,12',AlignedDim2='l,-2,2,12',AlignedDim3='E,-2,2,12')

# Create projection
proj_id = Projection([1,1,0], [-1,1,0], [0,0,1])
proj_id.setType(0,'r')
proj_id.setType(1,'r')
proj_id.setType(2,'r')
ws_id = proj_id.createWorkspace()

out_md_2d = CutMD(ws_mdew, Projection=ws_id, PBins=([-2,0.05,2],[-1.1,-0.9],[-0.1,0.1],[-2,0.05,2]), NoPix=True)

sv1=plotSlice(out_md_2d,colormin=0.01,colormax=1,normalization=2)
```
1. Ensure the non-integrated dimensions are selected in the slice viewer (the first and last dimension).
2. You should see something plotted other than only NaN for the slice.

Fixes #15263 .

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
